egs/librispeech/ASR/conformer_ctc/streaming_decode.py

```diff
@@ -213,11 +214,18 @@ def decode_one_batch(
     key = "ctc-greedy-search"
     batch_size = nnet_output.size(0)
     maxlen = nnet_output.size(1)
-    encoder_out_lens = torch.tensor([maxlen]).to(nnet_output.device)
     topk_prob, topk_index = nnet_output.topk(1, dim=2)  # (B, maxlen, 1)
     topk_index = topk_index.view(batch_size, maxlen)  # (B, maxlen)
     mask = make_pad_mask(encoder_out_lens).to(nnet_output.device)  # (B, maxlen)
-    topk_index = topk_index.masked_fill_(mask, eos_id)  # (B, maxlen)
+    # Though the padded place SHOULDB be filled with blank,
+    # it's even better without doing this,
```

Comment:
Yes, models are being tested and more results will posted here soon.

Besides not filling the padded place with blank, some experienced enginneers who familar this chunk based decoding methods, EVEN suggest padding more zeros to the raw pcm to make "the extracted feature" longer than "the real feature". 
They said this trick could help avoiding some deletion error at the end.
That's why I am padding some dummy tailing frames at the end of feature.
I am still tuning about this.

Reply:
This is probably happening because the model is delaying the output.  I am concerned that this delay may get longer, the longer we train.  It is worthwhile, i think, implementing that penalization think, I mentioned it by wechat (repeated below).  Not urgent, but we should probably add this at some point.

Guys (esp. Liyong), I just realized we have to be careful, when experimenting with models that have limited chunk size, that the network does not delay the symbols too much.  We should probably keep track of this somehow.  One metric could be the average time of all the symbols in the utterance (repeated just once, not with repeats), minus the midpoint of the utterance.   This could be interpreted as an "average seconds of delay", which we could compare with a CTC model that has infinite chunk size.

I believe Facebook or one other big company had a paper about how to fix this.  It should be quite easy to replicate in our framework: basically, after we get the lattice and before computing posteriors, do: "lattice.scores -= alpha * (lattice_arc_times - midpoint_time)", where the "lattice_arc_times" are frame indexes, that only appear on arcs that have symbols  on them (without repeats), which can be accomplished by a suitable product of two expressions, we won't have to write any C++ code.  The "- midpoint_time" won't affect the optimization, it's just to make the loss function more interpretable.
This will only work correctly if there is only one possible sequence of symbols for any given utterance; it wouldn't work with a lexicon that had optional silence or multiple pronunciations (i.e. only BPE).

Reply:
Just my 2c -- yeah padding with extra silence (250-500ms) at both sides of an utterance during decoding can sometimes really help get those WERs lower with a lot of models (unless there's silence there already). But IDK if people do that in training. I also like Dan's idea.

Reply:
> if people do that in training.

Now this trick is only used during decoding in my experiments.